cmd/snap-confine/mount-support.c

```diff
@@ -62,47 +62,82 @@
+	// persistent /tmp directory.
 	//
-	// Under that basedir, we put a 1777 /tmp dir that is then bind
-	// mounted for the applications to use
-	sc_must_snprintf(tmpdir, sizeof(tmpdir), "/tmp/snap.%s_XXXXXX", snap_name);
-	if (mkdtemp(tmpdir) == NULL) {
-		die("cannot create temporary directory essential for private /tmp");
+	char base_dir[MAX_BUF] = { 0 };
+	char tmp_dir[MAX_BUF] = { 0 };
+	int base_dir_fd SC_CLEANUP(sc_cleanup_close) = -1;
+	int tmp_dir_fd SC_CLEANUP(sc_cleanup_close) = -1;
+	sc_must_snprintf(tmp_dir, sizeof(tmp_dir), "/tmp/snap.%s/tmp",
+			 snap_name);
+	sc_must_snprintf(base_dir, sizeof(base_dir), "/tmp/snap.%s", snap_name);
+
+	// Create /tmp/snap.$SNAP_NAME/ 0700 root.root
+	if (mkdir(base_dir, 0700) < 0 && errno != EEXIST) {
+		die("cannot create base directory %s", base_dir);
 	}
-	// now we create a 1777 /tmp inside our private dir
-	mode_t old_mask = umask(0);
-	char *d = sc_strdup(tmpdir);
-	sc_must_snprintf(tmpdir, sizeof(tmpdir), "%s/tmp", d);
-	free(d);
-
-	if (mkdir(tmpdir, 01777) != 0) {
-		die("cannot create temporary directory for private /tmp");
+	base_dir_fd = open(base_dir,
+			   O_RDONLY | O_DIRECTORY | O_CLOEXEC | O_NOFOLLOW);
+	if (base_dir_fd < 0) {
+		die("cannot open base directory %s", base_dir);
 	}
-	umask(old_mask);
-
+	if (fchmod(base_dir_fd, 0700) < 0) {
+		die("cannot chmod base directory %s to 0700", base_dir);
+	}
+	if (fchown(base_dir_fd, 0, 0) < 0) {
+		die("cannot chown base directory %s to root.root", base_dir);
+	}
+	// Create /tmp/snap.$SNAP_NAME/tmp 01777 root.root
+	if (mkdirat(base_dir_fd, "tmp", 01777) < 0 && errno != EEXIST) {
+		die("cannot create private tmp directory %s/tmp", base_dir);
+	}
+	tmp_dir_fd = openat(base_dir_fd, "tmp",
+			    O_RDONLY | O_DIRECTORY | O_CLOEXEC | O_NOFOLLOW);
+	if (tmp_dir_fd < 0) {
+		die("cannot open private tmp directory %s/tmp", base_dir);
+	}
+	if (fchmod(tmp_dir_fd, 01777) < 0) {
+		die("cannot chmod private tmp directory %s/tmp to 01777",
+		    base_dir);
+	}
+	if (fchown(tmp_dir_fd, 0, 0) < 0) {
+		die("cannot chown private tmp directory %s/tmp to root.root",
+		    base_dir);
+	}
+	// TODO: move cwd jump to main of snap-confine.
 	// chdir to '/' since the mount won't apply to the current directory
-	char *pwd = get_current_dir_name();
-	if (pwd == NULL)
+	char *pwd SC_CLEANUP(sc_cleanup_string) = NULL;
+	pwd = get_current_dir_name();
+	if (pwd == NULL) {
 		die("cannot get current working directory");
-	if (chdir("/") != 0)
-		die("cannot change directory to '/'");
-
-	// MS_BIND is there from linux 2.4
-	sc_do_mount(tmpdir, "/tmp", NULL, MS_BIND, NULL);
-	// MS_PRIVATE needs linux > 2.6.11
+	}
+	if (chdir("/") < 0) {
+		die("cannot change directory to \"/\"");
+	}
+	sc_do_mount(tmp_dir, "/tmp", NULL, MS_BIND, NULL);
 	sc_do_mount("none", "/tmp", NULL, MS_PRIVATE, NULL);
-	// do the chown after the bind mount to avoid potential shenanigans
-	if (chown("/tmp/", 0, 0) < 0) {
-		die("cannot change ownership of /tmp");
```

Comment:
FYI, the chown was done after the mounts due to @tyhicks recommendation as part of 3f40f354757e6cb9dcfa37db57090796f4bdb6c2. It seems the history of the PR that introduced that is lost, but the idea is that before bdbfeebef03245176ae0dc323392bb0522a339b1, these directories could be created as non-root depending on the invoking user and moving the chown after making the dir private guarded against non-root doing things to the dir prior to the bind mount (and thus making the snap's /tmp have weird perms thereafter).

Today, we no longer chown to non-root so there is no potential for non-root to change the tmpdir to something else, so moving this chown earlier (and dropping the shenanigans comment) is safe.